0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,442 @@
+from Rust's). Note that this is different from the type of `not(true)`, which
+is `bool`. This difference is important to understanding higher-kindedness.
+
+In the analysis of kinds, all of these types - `bool`, `char`, `bool -> bool`
+and so on - have the kind `type`. Every type has the kind `type`. However,
+`type` is a base kind, just as `bool` is a base type, and there are terms with
+more complex kinds, such as `type -> type`. An example of a term of this kind
+is `Vec`, which takes a type as a parameter and evaluates to a type. The
+difference between the kind of `Vec` and the kind of `Vec<i32>` (which is
+`type`) is analogous to the difference between the type of `not` and
+`not(true)`. Note that `Vec<T>` has the kind `type`, just like `Vec<i32>`: even
+though `T` is a type parameter, `Vec` is still being applied to a type, just
+like `not(x)` still has the type `bool` even though `x` is a variable.
+
+A relatively uncommon feature of Rust is that it has _two_ base kinds, whereas
+many languages which deal with higher-kindedness only have the base kind
+`type`. The other base kind of Rust is the lifetime parameter. If you have a
+type like `Foo<'a>`, the kind of `Foo` is `lifetime -> type`.
+
+Higher-kinded terms can take multiple arguments as well, of course. `Result`
+has the kind `type, type -> type`. Given `vec::Iter<'a, T>` `vec::Iter` has the
+kind `lifetime, type -> type`.
+
+Terms of a higher kind are often called 'type operators'; the type operators
+which evaluate to a type are called 'type constructors'. There are other type
+operators which evaluate to other type operators, and there are even higher
+order type operators, which take type operators as their argument (so they have
+a kind like `(type -> type) -> type`). This RFC doesn't deal with anything as
+exotic as that.
+
+Specifically, the goal of this RFC is to allow type constructors to be
+associated with traits, just as you can currently associate functions, types,
+and consts with traits. There are other forms of polymorphism involving type
+constructors, such as implementing traits for a type constructor instead of a
+type, which are not a part of this RFC.
+
+## Features of associated type constructors
+
+### Declaring & assigning an associated type constructor
+
+This RFC proposes a very simple syntax for defining an associated type
+constructor, which looks a lot like the syntax for creating aliases for type
+constructors. The goal of using this syntax is to avoid to creating roadblocks
+for users who do not already understand higher kindedness.
+
+```rust
+trait StreamingIterator {
+   type Item<'a>;
+}
+```
+
+It is clear that the `Item` associated item is a type constructor, rather than
+a type, because it has a type parameter attached to it.
+
+Associated type constructors can be bounded, just like associated types can be:
+
+```rust
+trait Iterable {
+    type Item<'a>;
+    type Iter<'a>: Iterator<Item = Item<'a>>;
```

Comment:
Shouldn't this be `type Iter<'a>: Iterator<Item = Self::Item<'a>>;`? (Missing `Self::`)